File: sdks/dart/core/test/conversation_test.dart
import 'package:test/test.dart';
import 'package:vision_agents_core/vision_agents_core.dart';

/// Frames exactly as `frameOf` in the router writes them, so a change to the wire format on
/// that side fails here rather than in somebody's app.
AgentEvent frame(String json) => AgentEvent.tryParse(json)!;

void main() {
  group('Conversation', () {
    test('a reply arrives one delta at a time', () {
      var conversation = const Conversation();

      conversation = conversation.apply(
        frame('{"type":"responding","turn_id":"t1","prompt":"hi"}'),
      );
      expect(conversation.state, ConversationState.responding);
      expect(conversation.turns, hasLength(1));

      conversation = conversation
          .apply(frame('{"type":"response_delta","turn_id":"t1","text":"Hel"}'))
          .apply(frame('{"type":"response_delta","turn_id":"t1","text":"lo"}'));
      expect(conversation.turns.last.text, 'Hello');

      conversation = conversation.apply(
        frame(
          '{"pending_work":false,"type":"responded","turn_id":"t1","text":"Hello there.",'
          '"time_to_first_token_ms":90}',
        ),
      );
      expect(conversation.turns, hasLength(1));
      expect(conversation.turns.last.text, 'Hello there.');
      expect(conversation.turns.last.speaker.isAgent, isTrue);
      expect(conversation.state, ConversationState.idle);
    });

    test('keeps the turn id while the turn grows, so a list keyed by it keeps the row', () {
      final conversation = const Conversation()
          .apply(frame('{"type":"responding","turn_id":"t1"}'))
          .apply(frame('{"type":"response_delta","turn_id":"t1","text":"a"}'))
          .apply(frame('{"type":"response_delta","turn_id":"t1","text":"b"}'));

      expect(conversation.turns.map((turn) => turn.id), ['t1']);
    });

    test('a delta for a turn nobody saw begin still lands', () {
      final conversation = const Conversation().apply(
        frame('{"type":"response_delta","turn_id":"t9","text":"mid"}'),
      );

      expect(conversation.turns.map((turn) => turn.text), ['mid']);
      expect(conversation.turns.last.speaker.isAgent, isTrue);
    });

    test('a spoken turn with no final text keeps what was streamed', () {
      final conversation = const Conversation()
          .apply(frame('{"type":"responding","turn_id":"t1"}'))
          .apply(frame('{"type":"response_delta","turn_id":"t1","text":"Sure."}'))
          .apply(frame('{"type":"responded","turn_id":"t1","text":""}'));

      expect(conversation.turns.map((turn) => turn.text), ['Sure.']);
    });

    test('what was heard on a call becomes a participant turn', () {
      final conversation = const Conversation().apply(
        frame(
          '{"type":"heard","participant":{"id":"p1","user_id":"u1","name":"Alice"},'
          '"text":"what are your hours","language":"en"}',
        ),
      );

      expect(conversation.turns, hasLength(1));
      expect(conversation.turns.single.text, 'what are your hours');
      final speaker = conversation.turns.single.speaker;
      expect(speaker, isA<ParticipantSpeaker>());
      expect((speaker as ParticipantSpeaker).participant?.display, 'Alice');
    });

    test('typing something is not shown twice when the router echoes it', () {
      final conversation = const Conversation()
          .said('what are your hours')
          .apply(frame('{"type":"heard","participant":{},"text":"what are your hours"}'));

      expect(conversation.turns, hasLength(1));
    });

    test('a delegated skill is named while it runs', () {
      var conversation = const Conversation().apply(
        frame('{"type":"delegated","task_id":"k1","skill":"lookup_order"}'),
      );
      expect(conversation.state, const Working(['lookup_order']));

      conversation = conversation.apply(
        frame(
          '{"type":"task_settled","evidence":null,"task_id":"k1","skill":"lookup_order",'
          '"text":"done","question":"","elapsed_ms":12,"error":""}',
        ),
      );
      expect(conversation.state, ConversationState.responding);
    });

    test('two skills at once both show until both settle', () {
      var conversation = const Conversation()
          .apply(frame('{"type":"delegated","task_id":"k1","skill":"think"}'))
          .apply(frame('{"type":"delegated","task_id":"k2","skill":"recall"}'));
      expect(conversation.state, const Working(['think', 'recall']));

      conversation = conversation.apply(
        frame('{"type":"task_settled","task_id":"k1","skill":"think"}'),
      );
      expect(conversation.state, const Working(['recall']));
    });

    test('the conversation ends when the agent leaves', () {
      final conversation = const Conversation().apply(
        frame('{"type":"left","at":"2026-09-02T17:05:00Z"}'),
      );

      expect(conversation.state, ConversationState.ended);
    });

    test('a reported failure is kept rather than thrown', () {
      final conversation = const Conversation().apply(
        frame('{"type":"error","context":"tts","error":"the voice is unknown"}'),
      );

      expect(conversation.failure, 'the voice is unknown');
    });

    test('an event this SDK has never heard of changes nothing', () {
      final before = const Conversation().apply(frame('{"type":"responding","turn_id":"t1"}'));

      final after = before.apply(frame('{"type":"astonished","turn_id":"t1","degree":11}'));

      expect(identical(after, before), isTrue);
    });
  });

  group('AgentEvent', () {
    test('keeps an unknown event whole, with no kind and every field', () {
      final event = frame('{"type":"astonished","degree":11,"nested":{"a":[1,2]}}');

      expect(event.kind, isNull);
      expect(event.type, 'astonished');
      expect(event['degree'], 11);
      expect(event['nested'], {
        'a': [1, 2],
      });
    });

    test('reads nothing out of a frame that is not an object with a type', () {
      expect(AgentEvent.tryParse('not json'), isNull);
      expect(AgentEvent.tryParse('[1,2]'), isNull);
      expect(AgentEvent.tryParse('{"text":"no type"}'), isNull);
    });

    test('a tool call decodes its arguments, which arrive as a JSON string', () {
      final call = frame(
        '{"type":"tool_call","id":"c1","name":"lookup_order",'
        r'"arguments":"{\"order_id\":\"A-1042\"}","command_id":"","turn_id":"t1"}',
      ).toolCall!;

      expect(call.name, 'lookup_order');
      expect(call.argumentValues, {'order_id': 'A-1042'});
      expect(call.turnId, 't1');
    });

    test('arguments that are not an object decode to nothing rather than throwing', () {
      const call = ToolCall(id: 'c1', name: 'x', arguments: 'not json');

      expect(call.argumentValues, isEmpty);
    });
  });
}

File: sdks/kotlin/core/src/test/kotlin/io/getstream/visionagents/core/ConversationTest.kt
package io.getstream.visionagents.core

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertIs
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * Frames exactly as `frameOf` in the router writes them, so that a change to the wire format on
 * that side fails here rather than in somebody's app.
 */
private fun event(json: String): AgentEvent = requireNotNull(AgentEvent.decode(json))

private fun Conversation.after(vararg frames: String): Conversation =
    frames.fold(this) { conversation, frame -> conversation.reduce(event(frame)) }

class ConversationTest {
    @Test
    fun `a reply arrives one delta at a time and the final text wins`() {
        var conversation = Conversation().after("""{"type":"responding","turn_id":"t1","participant":{"id":"","user_id":"","name":""},"prompt":"hi"}""")
        assertEquals(Conversation.State.Responding, conversation.state)
        assertEquals(1, conversation.turns.size)

        conversation = conversation.after(
            """{"type":"response_delta","turn_id":"t1","text":"Hel"}""",
            """{"type":"response_delta","turn_id":"t1","text":"lo"}""",
        )
        assertEquals("Hello", conversation.turns.last().text)

        conversation = conversation.after(
            """{"pending_work":false,"type":"responded","turn_id":"t1","text":"Hello there.","time_to_first_token_ms":90}""",
        )
        assertEquals(listOf("Hello there."), conversation.turns.map { it.text })
        assertTrue(conversation.turns.last().isAgent)
        assertEquals(Conversation.State.Idle, conversation.state)
    }

    @Test
    fun `a delta for a turn nobody saw begin still lands`() {
        val conversation = Conversation().after("""{"type":"response_delta","turn_id":"t9","text":"mid"}""")

        assertEquals(listOf("mid"), conversation.turns.map { it.text })
        assertTrue(conversation.turns.last().isAgent)
    }

    @Test
    fun `a spoken turn with no final text keeps what was streamed`() {
        val conversation = Conversation().after(
            """{"type":"responding","turn_id":"t1"}""",
            """{"type":"response_delta","turn_id":"t1","text":"Sure."}""",
            """{"type":"responded","turn_id":"t1","text":""}""",
        )

        assertEquals(listOf("Sure."), conversation.turns.map { it.text })
    }

    @Test
    fun `what was heard on a call becomes a participant turn`() {
        val conversation = Conversation().after(
            """{"type":"heard","participant":{"id":"p1","user_id":"u1","name":"Alice"},"text":"what are your hours","language":"en"}""",
        )

        val turn = conversation.turns.single()
        assertEquals("what are your hours", turn.text)
        val speaker = assertIs<Turn.Speaker.Person>(turn.speaker)
        assertEquals("Alice", speaker.participant?.display)
    }

    @Test
    fun `typing something is not shown twice when the router echoes it`() {
        val conversation = Conversation()
            .said("what are your hours")
            .after("""{"type":"heard","participant":{},"text":"what are your hours"}""")

        assertEquals(1, conversation.turns.size)
    }

    @Test
    fun `a delegated skill is named while it runs`() {
        var conversation = Conversation().after("""{"type":"delegated","task_id":"k1","skill":"lookup_order","prompt":"","turn_id":"t1"}""")
        assertEquals(Conversation.State.Working(listOf("lookup_order")), conversation.state)

        conversation = conversation.after(
            """{"type":"task_settled","evidence":null,"task_id":"k1","skill":"lookup_order","text":"done","question":"","elapsed_ms":12,"error":""}""",
        )
        assertEquals(Conversation.State.Responding, conversation.state)
    }

    @Test
    fun `two skills at once both show until both settle`() {
        val conversation = Conversation().after(
            """{"type":"delegated","task_id":"k1","skill":"think"}""",
            """{"type":"delegated","task_id":"k2","skill":"recall"}""",
            """{"type":"task_settled","task_id":"k1","skill":"think"}""",
        )

        assertEquals(Conversation.State.Working(listOf("recall")), conversation.state)
    }

    @Test
    fun `the conversation ends when the agent leaves`() {
        val conversation = Conversation().after("""{"type":"left","at":"2026-09-02T17:05:00Z"}""")

        assertEquals(Conversation.State.Ended, conversation.state)
    }

    @Test
    fun `a reported failure is kept rather than thrown`() {
        val conversation = Conversation().after("""{"type":"error","context":"tts","error":"the voice is unknown"}""")

        assertEquals("the voice is unknown", conversation.failure)
    }

    @Test
    fun `an interruption leaves the agent waiting to be spoken to`() {
        val conversation = Conversation().after(
            """{"type":"responding","turn_id":"t1"}""",
            """{"type":"interrupted","turn_id":"t1","participant":{"id":"p1","user_id":"u1","name":""}}""",
        )

        assertEquals(Conversation.State.Idle, conversation.state)
    }

    @Test
    fun `an event this SDK has never heard of changes nothing`() {
        val before = Conversation().after("""{"type":"responding","turn_id":"t1"}""")

        val after = before.after("""{"type":"astonished","turn_id":"t1","degree":11}""")

        assertEquals(before, after)
        assertNull(event("""{"type":"astonished"}""").kind)
    }
}
